discovery/discovery.go

```diff
@@ -221,7 +221,7 @@ func (d *discovery) checkCluster() ([]*client.Node, int, uint64, error) {
 		if err == client.ErrInvalidJSON {
 			return nil, 0, 0, ErrBadDiscoveryEndpoint
 		}
-		if err == context.DeadlineExceeded {
+		if _, ok := err.(*client.ClusterError); ok {
```

Comment:
> If we decide to retry forever, we should retry forever.

Make sense. We have decided to retry forever for connection error. I still think that it needs to print out detailed error to help user understand why it failed.